src/gap_to_julia.jl

```diff
@@ -8,24 +7,60 @@ end
 Base.showerror(io::IO, e::ConversionError) =
     print(io, "failed to convert $(typeof(e.obj)) to $(e.jl_type):\n $(e.obj)")
 
+
 """
     RecDict
 
-An internal type of GAP.jl used for tracking conversion results in `gap_to_julia`.
+An internal type of GAP.jl used for tracking conversion results in `gap_to_julia_intern`.
+"""
+const RecDict = IdDict{Tuple{Type,Any},Any}
+
+"""
+    gap_to_julia_intern(type::Type, input::Any, recursive::Bool, dict::Union{Nothing,RecDict})
+
+FIXME: update this text
+
+Try to convert the object `x` to a Julia object of type `type`.
+If `x` is a `GAP.GapObj` then the conversion rules are defined in the
+manual of the GAP package JuliaInterface.
+If `x` is another `GAP.Obj` (for example a `Int64`) then the result is
+defined in Julia by `type`.
+
+The parameter `dict` is used to preserve the identity
+of converted subobjects.
+
+For GAP lists and records, it makes sense to convert also the subobjects
+recursively, or to keep the subobjects as they are;
+the behaviour is controlled by `recursive`, which can be `true` or `false`.
 """
-const RecDict = IdDict{Any,Any}
+function gap_to_julia_intern end
 
 ## Conversion from GAP to Julia
 """
-    gap_to_julia(type, x, recursion_dict::Union{Nothing,RecDict}=nothing; recursive::Bool=true)
+    gap_to_julia(type::Type, x::Any; recursive::Bool = true)
+
+FIXME: update this text
+TODO: document that `type` must be a `Any` or a concrete type, and likewise
+for its type parameters. This means that `Tuple{Any}` is not allowed, nor
+are unions. Indeed, for a union, how should we determine which of the various
+union types should be the end result of the conversion?
```

Comment:
I do not understand this statement.
Currently we have
```
julia> GAP.gap_to_julia(GAP.evalstr("[1,Z(2),true]"))
3-element Vector{Any}:
    1
     GAP: Z(2)^0
 true
```
and
```
julia> Vector{GAP.Obj}(GAP.evalstr("[1,Z(2),true]"))
3-element Vector{Union{Bool, GAP.FFE, GAP.GapInt}}:
    1
     GAP: Z(2)^0
 true
```
Shall the latter be forbidden?

Reply:
So we already discussed this on gather, but I want to record what I said here for others / myself in the future:

Indeed the text is not phrased correctly. What I wanted to express is that I don't see how to handle a request like this properly:
`Vector{Union{String,Int,Symbol}}(GAP.evalstr("[1, \"x\"]"))`

In principle this could be dealt with, in that there are sensible conversion results of the requested type for this, namely `[1, "x"]` and `[1, :x]`. But then we need to decide which of them to return? Perhaps do it arbitrarily? Or introduce some kind of priority between certain types (e.g "if both `Symbol` and `String` are requested and possible, prefer `Symbol`"), but then how to extend that to all cases (We could be asked to produce a `Vector{Union{Vector{Vector{Int}}, Matrix{Int}, fmpz_mat, fmpq_mat}` ...).

In contrast, conversion to `Vector{GAP.Obj}`  above works, despite `GAP.Obj` being a union type, because all the elements of the GAP list already had a type in that Union. So in that case the "trivial conversions" rule applies, i.e., asking to convert an object `x::T` to type `S` always succeeds and returns `x` whenever `S` is either `T` or a supertype of `T`.